include/pagmo/utils/gradients_and_hessians.hpp

```diff
@@ -0,0 +1,247 @@
+/* Copyright 2017 PaGMO development team
+
+This file is part of the PaGMO library.
+
+The PaGMO library is free software; you can redistribute it and/or modify
+it under the terms of either:
+
+  * the GNU Lesser General Public License as published by the Free
+    Software Foundation; either version 3 of the License, or (at your
+    option) any later version.
+
+or
+
+  * the GNU General Public License as published by the Free Software
+    Foundation; either version 3 of the License, or (at your option) any
+    later version.
+
+or both in parallel, as here.
+
+The PaGMO library is distributed in the hope that it will be useful, but
+WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+You should have received copies of the GNU General Public License and the
+GNU Lesser General Public License along with the PaGMO library.  If not,
+see https://www.gnu.org/licenses/. */
+
+#ifndef PAGMO_UTILS_GRADIENTS_AND_HESSIANS_HPP
+#define PAGMO_UTILS_GRADIENTS_AND_HESSIANS_HPP
+
+/** \file gradients_and_hessians.hpp
+ * \brief Utilities of general interest for gradients and hessians related calculations
+ *
+ * This header contains utilities useful in general for gradients and hessians related calculations
+ */
+
+#include <algorithm>
+#include <cmath>
+#include <stdexcept>
+#include <vector>
+
+#include "../exceptions.hpp"
+#include "../problem.hpp"
+#include "../types.hpp"
+
+namespace pagmo
+{
+/// Heuristic to estimate the sparsity pattern
+/**
+ * A numerical estimation of the sparsity pattern of same callable object is made by numerically
+ * computing it around a given input point and detecting the components that are changed.
+ *
+ * The callable function \p f must have the prototype:
+ *
+ * @code{.unparsed}
+ * vector_double f(vector_double x)
+ * @endcode
+ *
+ * otherwise compiler errors will be generated.
+ *
+ * Note that estimate_sparsity may fail to detect the real sparsity as it only considers one variation around the input
+ * point. It is of use, though, in tests or cases where its not possible to write the sparsity or where the user is
+ * confident the estimate will be correct.
+ *
+ * @tparam Func a callable object assumed to be returning a fitness vector when called on \p x
```

Comment:
I would drop this ``tparam``, as we are not using this command anywhere else and you already have described the reqs on ``Func`` above.